File: 2-rate-limiting-with-queues/0-typescript/src/rate-limiter.ts
import { readFileSync } from "fs"
import { join } from "path"
import type { Redis } from "ioredis"
import type { AppConfig } from "./config"

export interface RateCheckResult {
  readonly allowed: boolean
  readonly remaining: number
  readonly retryAfterMs: number
}

// Edge-layer token bucket. The whole check-then-decrement runs as ONE atomic
// Lua script in Redis (EVALSHA), so concurrent requests cannot interleave.
export class RateLimiter {
  private luaSha = ""
  private readonly luaScript: string

  constructor(
    private readonly redis: Redis,
    private readonly config: AppConfig,
  ) {
    const scriptPath = join(__dirname, "lua", "token-bucket.lua")
    this.luaScript = readFileSync(scriptPath, "utf-8")
  }

  // Preload the script so EVALSHA can reuse the cached SHA on every call.
  async init(): Promise<void> {
    this.luaSha = (await this.redis.script("LOAD", this.luaScript)) as string
  }

  get scriptSha(): string {
    return this.luaSha
  }

  // One atomic round-trip: returns { allowed, remaining, retryAfterMs }.
  async checkRate(clientKey: string): Promise<RateCheckResult> {
    const now = Date.now()
    const key = `rl:${clientKey}`

    const result = (await this.redis.evalsha(
      this.luaSha,
      1,
      key,
      this.config.rateCapacity,
      this.config.rateRefillPerSec,
      now,
      1,
    )) as [number, number, number]

    return {
      allowed: result[0] === 1,
      remaining: result[1],
      retryAfterMs: result[2],
    }
  }
}

File: 2-rate-limiting-with-queues/2-csharp/RateLimiter.cs
using StackExchange.Redis;

namespace RateLimitApi;

public readonly record struct RateCheckResult(bool Allowed, long Remaining, long RetryAfterMs);

/// <summary>
/// Edge-layer token bucket. The whole check-then-decrement runs as ONE atomic
/// Lua script in Redis (EVALSHA), so concurrent requests cannot interleave.
/// </summary>
public sealed class RateLimiter
{
    private readonly IDatabase _redis;
    private readonly int _capacity;
    private readonly int _refillPerSec;
    private byte[] _luaSha = Array.Empty<byte>();
    private readonly string _luaScript;

    public RateLimiter(IConnectionMultiplexer mux, IConfiguration config)
    {
        _redis = mux.GetDatabase();
        _capacity = config.GetValue("Rate:Capacity", 5);
        _refillPerSec = config.GetValue("Rate:RefillPerSec", 5);
        _luaScript = File.ReadAllText(Path.Combine(AppContext.BaseDirectory, "token-bucket.lua"));
    }

    // Preload the script so later calls send only the cached SHA.
    public async Task InitAsync(IConnectionMultiplexer mux)
    {
        var endpoint = mux.GetEndPoints()[0];
        _luaSha = (byte[])(await mux.GetServer(endpoint).ScriptLoadAsync(_luaScript))!;
    }

    // One atomic round-trip: returns (allowed, remaining, retryAfterMs).
    public async Task<RateCheckResult> CheckRateAsync(string clientKey)
    {
        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        var key = $"rl:{clientKey}";

        var raw = (RedisValue[])(await _redis.ScriptEvaluateAsync(
            _luaSha,
            new RedisKey[] { key },
            new RedisValue[] { _capacity, _refillPerSec, now, 1 }))!;

        return new RateCheckResult((long)raw[0] == 1, (long)raw[1], (long)raw[2]);
    }
}
